File: programmers/level2/피로도.py
answer = 0


def solution(k, dungeons):
    def DFS(L):
        global answer
        if L == len(dungeons):
            temp = k
            cnt = 0
            for x in res:
                if temp < dungeons[x][0]:
                    break
                else:
                    temp -= dungeons[x][1]
                    cnt += 1
            if cnt > answer:
                answer = cnt
            return
        else:
            for i in range(len(dungeons)):
                if ch[i] == 0:
                    ch[i] = 1
                    res[L] = i
                    DFS(L + 1)
                    ch[i] = 0

    ch = [0 for _ in range(len(dungeons))]
    res = [0 for _ in range(len(dungeons))]
    DFS(0)
    return answer


print(solution(80, [[80,20],[50,40],[30,10]]))

File: 프로그래머스/2/87946. 피로도/피로도.java
//dfs로 순서가 있는 배열을 만듦
//배열의 사이즈가 dungeons와 같으면 처리 시작

import java.util.*;

class Solution {
    static int ans = Integer.MIN_VALUE;
    static boolean[] visited;
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        DFS(k, dungeons, new ArrayList<int[]>());
        return ans;
    }
    
    private void DFS(int k, int[][] dungeons, List<int[]> dungeon) {
        if(dungeons.length==dungeon.size()) {
            int cnt = 0;
            for(int[] arr:dungeon) {
                //최소 필요 소모도가 k보다 작거나 같고, 소모 피로도가 k보다 작거나 같으면
                if(arr[0]<=k && arr[1]<=k) {
                    cnt++;
                    k -= arr[1];
                }
            }
            ans = Math.max(ans, cnt);
        } else {
            for(int i=0; i<dungeons.length; i++) {
                if(!visited[i]) {
                    visited[i] = true;
                    dungeon.add(dungeons[i]);
                    DFS(k, dungeons, dungeon);
                    visited[i] = false;
                    dungeon.remove(dungeons[i]);
                }
            }
        }
    }
}